prototype2/gdgem/nmx/Readout.h

```diff
@@ -1,40 +1,98 @@
 /* Copyright (C) 2017-2018 European Spallation Source, ERIC. See LICENSE file */
 //===----------------------------------------------------------------------===//
 ///
+///                              WARNING
+///
+///                              ACHTUNG
+///
+///                              AVISO
+///
+///                              ADVARSEL
+///
+///                              DĖMESIO
+///
+///                              UWAGA
+///
+///
+///
+///          MODIFY THIS FILE ONLY AFTER UNDERSTANDING THE FOLLOWING
+///
+///
+///   Any changes to non-static variable definitions will likely break h5 file
+/// writing and compatibility. If you rename, reorder or change the type of any
+/// of the member variables, you MUST:
+///    A) Increment FormatVersion by 1
+///    B) Ensure the hdf5 TypeTrait maps the struct correctly
+///
+/// If you cannot ensure the above, consult someone who can.
+///
 /// \file
 ///
+/// \brief Readout struct for NMX event formation
+///
 //===----------------------------------------------------------------------===//
 
 #pragma once
 
-#include <cinttypes>
-
-// TODO: initialize values to 0?
-struct Readout
-{
-  uint8_t fec;
-  uint8_t chip_id;
-  uint32_t bonus_timestamp;
-  uint32_t srs_timestamp;
-  uint16_t channel;
-  uint16_t bcid;
-  uint16_t tdc;
-  uint16_t adc;
-  float ChipTimeNs;
-  bool over_threshold;
-
-  bool operator==(const Readout& other) const
-  {
+#include <common/DumpFile.h>
+
+struct __attribute__ ((packed)) Readout {
+  // \todo use constexpr string_view when c++17 arrives
+  static std::string DatasetName() { return "srs_hits"; }
+  static uint16_t FormatVersion() { return 0; }
+
+  // !!! DO NOT MODIFY BELOW - READ HEADER FIRST !!!
```

Comment:
Worth reordering for memory use?

Reply:
Elaborate, please? :)

Reply:
https://jonasdevlieghere.com/order-your-members/

Reply:
I would argue against Matt here. If you have not done any profiling, it is very likely that optimising the structure is wasted effort. If you have a performance problem and profiling of the code shows that memory access is a bottleneck, do optimise.

Reply:
@SkyToGround It would be for memory use rather than CPU time. I ask the question "is worth it?" because I don't know how many of the objects are likely to exist at a given time. Also the "wasted effort" is near zero... reordering some members.

Reply:
RAM is super cheap these days so unless you can show that saving a few bytes here and there is worth it, optimising for memory usage is in most cases a waste of time in my opinion. I would also argue that the effort is not near zero as doing the re-ordering requires some mental capacity and more importantly, time.

Reply:
Sure, it may not be worth it in this case. As I mentioned, it depends how many of these objects exist. I think it is a good thing to be aware of and be in the habit of doing in general though.

Reply:
I second Matt. And then I would add that designing code for efficiency or laying out data structures sensibly from the beginning is not optimisation. But that is another discussion...